x-pack/plugins/enterprise_search/public/applications/workplace_search/views/search_authorize/search_authorize_logic.ts

```diff
@@ -0,0 +1,119 @@
+
+interface SearchAuthorizeValues {
+  redirectPending: boolean;
+  cachedPreAuth: OAuthPreAuthorization;
+}
+
+interface SearchAuthorizeActions {
+  setServerProps(serverProps: OAuthPreAuthServerProps): OAuthPreAuthServerProps;
+  initializeSearchAuth(searchOAuth: SearchOAuth): SearchOAuth;
+  authorizeSearch(): void;
+  setRedirectNotPending(): void;
+}
+
+export const SearchAuthorizeLogic = kea<
+  MakeLogicType<SearchAuthorizeValues, SearchAuthorizeActions>
+>({
+  path: ['enterprise_search', 'workplace_search', 'search_authorize_logic'],
+  actions: {
+    setServerProps: (serverProps: OAuthPreAuthServerProps) => serverProps,
+    initializeSearchAuth: (searchOAuth: SearchOAuth) => searchOAuth,
+    authorizeSearch: () => null,
+    setRedirectNotPending: () => null,
+  },
+  reducers: {
+    redirectPending: [
+      true,
+      {
+        setRedirectNotPending: () => false,
+      },
+    ],
+    cachedPreAuth: [
+      {} as OAuthPreAuthorization,
+      {
+        setServerProps: (_, serverProps) => transformServerPreAuth(serverProps),
+      },
+    ],
+  },
+  listeners: ({ actions, values }) => ({
+    initializeSearchAuth: async (searchOAuth) => {
+      clearFlashMessages();
+      const { http } = HttpLogic.values;
+
+      const query = {
+        client_id: searchOAuth.clientId,
+        response_type: 'code',
+        redirect_uri: searchOAuth.redirectUrl,
+        scope: 'search',
+      };
+
+      try {
+        const response = await http.get(oauthAuthorizeRoute, { query });
+
+        if (response.status === 'redirect') {
+          window.location.replace(response.redirect_uri);
+        } else {
+          actions.setServerProps(response);
+        }
+      } catch (e) {
+        flashAPIErrors(e);
+        actions.setRedirectNotPending();
+      }
+    },
+    setServerProps: () => {
+      actions.authorizeSearch();
+    },
+    authorizeSearch: async () => {
+      const { http } = HttpLogic.values;
+      const { cachedPreAuth } = values;
+
+      try {
+        const response = await http.post(oauthAuthorizeRoute, {
+          body: JSON.stringify({
+            client_id: cachedPreAuth.clientId,
+            response_type: cachedPreAuth.responseType,
+            redirect_uri: cachedPreAuth.redirectUri,
+            scope: cachedPreAuth.rawScopes,
+          }),
+          headers,
+        });
+
+        window.location.replace(response.redirect_uri);
+      } catch (e) {
+        flashAPIErrors(e);
+        actions.setRedirectNotPending();
+      }
+    },
+  }),
+});
+
+const transformServerPreAuth = (serverProps: OAuthPreAuthServerProps): OAuthPreAuthorization => {
+  return {
+    clientId: serverProps.client_id,
+    clientName: serverProps.client_name,
+    redirectUri: serverProps.redirect_uri,
+    responseType: serverProps.response_type,
+    rawScopes: serverProps.scope,
+    scopes: serverProps.scope.split(', '),
+    state: serverProps.state,
+  };
+};
```

Comment:
Might be best to DRY this out to a util since it's used in  `oauth_authorize_logic.ts`


Coudld be: `x-pack/plugins/enterprise_search/public/applications/workplace_search/views/oauth_authorize/utils/transform_server_pre_auth.ts`

Reply:
I'm thinking I'll just export it from `oauth_authorize_logic`, since it uses interfaces that are already exported from `oauth_authorize_logic` to `search_authorize_logic`.